satpy/modifiers/parallax.py

```diff
@@ -0,0 +1,537 @@
+# terms of the GNU General Public License as published by the Free Software
+# Foundation, either version 3 of the License, or (at your option) any later
+# version.
+#
+# satpy is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# satpy.  If not, see <http://www.gnu.org/licenses/>.
+"""Parallax correction.
+
+Routines related to parallax correction using datasets involving height, such
+as cloud top height.
+
+The geolocation of (geostationary) satellite imagery is calculated by
+agencies or in satpy readers with the assumption of a clear view from
+the satellite to the geoid.  When a cloud blocks the view of the Earth
+surface or the surface is above sea level, the geolocation is not accurate
+for the cloud or mountain top.  This module contains routines to correct
+imagery such that pixels are shifted or interpolated to correct for this
+parallax effect.
+
+Parallax correction is currently only supported for (cloud top) height
+that arrives on an :class:`~pyresample.geometry.AreaDefinition`, such
+as is standard for geostationary satellites.  Parallax correction with
+data described by a :class:`~pyresample.geometry.SwathDefinition`,
+such as is common for polar satellites, is not (yet) supported.
+
+See also the :doc:`../modifiers` page in the documentation for an introduction to
+parallax correction as a modifier in Satpy.
+"""
+
+import datetime
+import inspect
+import logging
+import warnings
+
+import dask.array as da
+import numpy as np
+import xarray as xr
+from pyorbital.orbital import A as EARTH_RADIUS
+from pyorbital.orbital import get_observer_look
+from pyresample.bucket import BucketResampler
+from pyresample.geometry import SwathDefinition
+
+from satpy.modifiers import ModifierBase
+from satpy.resample import resample_dataset
+from satpy.utils import get_satpos, lonlat2xyz, xyz2lonlat
+
+logger = logging.getLogger(__name__)
+
+
+class MissingHeightError(ValueError):
+    """Raised when heights do not overlap with area to be corrected."""
+
+
+class IncompleteHeightWarning(UserWarning):
+    """Raised when heights only partially overlap with area to be corrected."""
+
+
+def forward_parallax(sat_lon, sat_lat, sat_alt, lon, lat, height):
+    """Calculate forward parallax effect.
+
+    Calculate the forward parallax effect.  When a satellite instrument
+    observes the Earth, the geolocation assumes it sees the Earth surface
+    at the geoid (elevation zero).  In reality, the field of view may
+    stop short of the geoid as it observes, for example, a cloud or
+    elevated ground.  This function calculates the forward parallax
+    effect.  If the view of a pixel at location (lat, lon) is blocked
+    by a cloud at height h, we calculate the location of this blocking.
```

Comment:
Maybe I would rewrite it as follow: 
`If the view of a pixel at location (lat, lon) is blocked by a cloud at height h, this function calculates the (lat, lon) coordinates of an observer located at the Earth surface below the sight-blocking cloud.`


    Calculate parallax correction based on satellite position and
    (cloud top) height coordinates in geodetic (unprojected) coordinates.
    This function calculates the latitude and longitude belonging to the
    cloud top, based on the location of the satellite and the location
    of the cloud.